### ISSUE TITLE
bug with height encoding in TimeSeries DSG?

### ISSUE THREAD

**Issue by emiliom** [ioos/i52n-sos#24]
I think there's an important bug with how the IOOS 52N SOS presents heights in TimeSeries DSG data. I'll start with the caveat that I've only looked at `0.8.1-SNAPSHOT` GetObservation responses, but I don't remember seeing anything about relevant enhancements in 1.0.0. Another caveat may be that it's unclear whether this is a problem with the 52North server software, or with `sos-injector`; I supect it's the former, though.

I've looked closely at both single-station and multi-station TimeSeries responses. I've examined two responses from the NANOOS i52n SOS and one from the GCOOS i52n SOS (both of which are on `0.8.1-SNAPSHOT` and use `sos-injector`; pinging @fgayanilo b/c I refer a lot to his GCOOS SOS!). For reference, here are two sample `GetObservation` requests, one from each server (adjust `eventTime` as you see fit):

- [NANOOS sea water temperature @ `urn:ioos:station:nanoos:psi_baycenter`](http://data.nanoos.org/52nsos/sos/kvp?service=SOS&version=1.0.0&request=GetObservation&responseFormat=text/xml; subtype="om/1.0.0/profiles/ioos_sos/1.0"&offering=urn:ioos:station:nanoos:psi_baycenter&observedProperty=http://mmisw.org/ont/cf/parameter/sea_water_temperature&eventTime=2016-1-3T00:00Z/2016-1-5T23:00Z)
- [GCOOS sea water temperature @ `urn:ioos:station::tcoon:072`]( http://data.gcoos.org:8080/52nSOS/sos/kvp?service=SOS&version=1.0.0&request=GetObservation&responseFormat=text/xml; subtype="om/1.0.0/profiles/ioos_sos/1.0"&offering=urn:ioos:station:tcoon:072&observedProperty=http://mmisw.org/ont/cf/parameter/sea_water_temperature&eventTime=2016-1-3T00:00Z/2016-1-5T23:00Z)

The SWE `height` elements never include any height values. Specifically, see these XPaths, both relative to `/om:ObservationCollection/om:member/om:Observation/om:result/swe2:DataRecord/swe2:field[@name=stations]`:
- `swe2:DataRecord/swe2:field/swe2:DataRecord/swe2:field[@name=platformLocation]/swe2:Vector/swe2:coordinate[@name=height]`
- `swe2:DataRecord/swe2:field/swe2:DataRecord/swe2:field[@name=sensors]/swe2:DataRecord/swe2:field/swe2:DataRecord/swe2:field[@name=height]`

Instead, height values are stored in `profileHeights` elements, which don't belong in a TimeSeries DSG response. For reference, see the IOOS SOS Guidelines and templates. eg:
- http://ioos.github.io/sos-guidelines/doc/wsdd/sos_wsdd_github_notoc/#getobservation-response
- https://github.com/ioos/sos-guidelines/blob/master/template/milestone1.0/SWE-MultiStation-TimeSeries.xml#L195
- https://github.com/ioos/sos-guidelines/blob/master/template/milestone1.0/SWE-MultiStation-TimeSeries.xml#L222

FYI, I also looked closely at the [pyoos `IoosSweSos` implementation of TimeSeries](https://github.com/ioos/pyoos/blob/master/pyoos/parsers/ioos/one/timeseries.py), to be extra sure. I took one of the NANOOS SOS GetObservation responses and tweaked it manually to match the Milestone 1.0 template expectation, and sure enough, pyoos could parse it correctly (with one or two caveats involving two likely pyoos bugs, one of which I'm ready to fix).

If it helps, I can upload to a github repo or gists my original sample GetObservation responses.

**Comment by abirger** [comment]
@emiliom , ncSOS reports height value for `timeSeries` in a right way. For example, see response to that [GetObservation request](http://sos.maracoos.org/pre/sos/wflow700-agg.ncml?service=SOS&version=1.0.0&request=GetObservation&offering=urn:ioos:station:ncsos:station-700&observedProperty=http://mmisw.org/ont/cf/parameter/wind_speed&eventtime=2014-03-02T00:58:00Z/2014-03-02T03:08:00Z&responseFormat=text%2Fxml%3Bsubtype%3D%22om%2F1.0.0%2Fprofiles%2Fioos_sos%2F1.0%22).

You may see that it reports height value where it belongs, i.e. at XPATH
```
/om:ObservationCollection/om:member[1]/om:Observation[1]/om:result[1]/swe2:DataRecord[1]/swe2:field[1]/swe2:DataRecord[1]/swe2:field[1]/swe2:DataRecord[1]/swe2:field[2]/swe2:Vector[1]/swe2:coordinate[3]/@name
```

**Comment by emiliom** [comment]
Thanks Alex! I had not bothered to look at ncsos responses. It's really
helpful to have that confirmation here. Plus, it'll help me as I
investigate and fix possible pyoos bugs with height handling.

**Comment by emiliom** [comment]
@abirger: when I first saw your posting on my smartphone I clicked on your GetObservation link and got a realistic-looking response (as tiny as everything looked on a phone). Now, three hours later, I tried clicking on it again both on my desktop browser and on my phone, but I'm always getting an `ows:ExceptionReport` (`Failed to open dataset &lt;/wflow700-agg.ncml&gt;: java.lang.NullPointerException`). Weird!

If you or anyone else could post a more reliable ncSOS sample GetObservation request, that'd be much appreciated.

**Comment by kwilcox** [comment]
@lukecampbell see above for tds error

**Comment by lukecampbell** [comment]
We're restarting and upgrading the sos.maracoos.org instance. I just clicked the link a few minutes ago and it worked for me.

**Comment by abirger** [comment]
@emiliom , although the ncSOS seems to be OK now, I've put the GetObservation response on [Google Drive](https://drive.google.com/a/noaa.gov/file/d/0ByFYUQ5Z5zQLLVZaZDhXeTV1LXc/view?usp=sharing) to preserve it just in case.

**Comment by emiliom** [comment]
Thanks @lukecampbell for bringing that server back up. It's working for me too. @abirger, I can't access the copy you put on Google Drive, but no worries. I've downloaded a response. Next week I'll probably upload it and a few other sample responses to a personal github repo for sharing and testing.

**Comment by emiliom** [comment]
I've gone ahead and created said repo: https://github.com/emiliom/stuff/tree/master/ioossosxml  Currently it has 4 GetObs responses. I'll add a very brief README later; in the meantime, the commit description for each file will give most of the background one would need.

But I guess I/we are starting to diverge from the i52n-sos issue at hand. It seems very clear now that 52N SOS does have a bug with height encoding in TimeSeries (@abirger, can you confirm?). I created the responses repo for interim documentation and to help me test the pyoos parser.

**Comment by abirger** [comment]
@emiliom , I can definitely confirm that `profileHeights` does not belong to a `timeSeries` template but is rather a part of a `timeSeriesProfile` one. So, NANOOS and GCOOS either misuse the `profileHeights` element or wrongly describe observation type in `/om:ObservationCollection/om:member[1]/om:Observation[1]/gml:description[1]`. I cannot point to the source of the error, let's wait for Shane's comment.

**Comment by emiliom** [comment]
The plot thickens, or clarifies a bit. Here are additional comments from @abirger (sent to me via email):

> Attached are the GetObs from GLOS and CeNCOOS that I refreshed today for your GitHub collection. They indeed do not report sensor height value under `swe2:DataRecord/swe2:field/swe2:DataRecord/swe2:field[@name=sensors]/swe2:DataRecord/swe2:field/swe2:DataRecord/swe2:field[@name=height]` but they do not insert a `profileHeights` element either. It seems to be specific for NANOOS & GCOOS.

I've [uploaded to my repo](https://github.com/emiliom/stuff/tree/master/ioossosxml) the 3 GetObs responses @abirger sent me. We now have 6 52N SOS samples there plus one ncSOS sample. Nice!

**Comment by abirger** [comment]
@emiliom , if you want, I can send you more samples. I have tons of them collected through the last couple of years while I was running the tests. Some of them are totally outdated but I believe that the samples from 2015 would be still viable.

**Comment by emiliom** [comment]
Based on @abirger's latest findings from GLOS and CeNCOOS 52N responses, my current guess is that there may be some interaction between `sos-injector` (which I know both GCOOS and NANOOS use) and the 52n SOS server per se. 

FYI, GLOS is using `0.8.0-SNAPSHOT` and CeNCOOS `1.0.0`.

Here's where things stand:
- We have not found a single example of a 52N SOS GetObs response that includes height values in the right XPaths (for `platformLocation` and optionally `sensors`).
- `profileHeights` are found in 2 (GCOOS and NANOOS) of the 4 52N SOS server instances. Those two use `sos-injector`, but I don't know what GLOS and CeNCOOS use. Also, of the 6 52N SOS examples, only the ones that have `profileHeights` elements actually have height values anywhere (but only in the `profileHeights` elements, unfortunately).

My *guess assessment* is that GLOS and CeNCOOS have not added z/height information to their assets, while GCOOS and NANOOS have; and the 52N server does have a bug for `timeSeries`, such that when z/height values are present, they're only presented via `profileHeights` elements; when no z/height values are present, `profileHeights` elements are not created.

**Comment by emiliom** [comment]
> if you want, I can send you more samples

Thanks for the offer, @abirger. But I think we have enough, for now. Let's wait to hear from @shane-axiom

**Comment by shane-axiom** [comment]
Ok, nice catch and thorough documentation @emiliom. I implemented some fixes for this in commit 4ee736c over the weekend. Example (test data) output for both timeSeries and timeSeriesProfile can be seen here:

http://ioossos.axiomalaska.com/i52n-sos-dev/service?service=SOS&version=1.0.0&request=GetObservation&offering=urn%3Aioos%3Anetwork%3Atest%3Aall&observedProperty=http%3A%2F%2Fmmisw.org%2Font%2Fcf%2Fparameter%2Fair_temperature,http%3A%2F%2Fmmisw.org%2Font%2Fcf%2Fparameter%2Fsea_water_temperature&procedure=urn%3Aioos%3Astation%3Atest%3A1&responseFormat=text%2Fxml%3B%20subtype%3D%22om%2F1.0.0%2Fprofiles%2Fioos_sos%2F1.0%22

This fixes the issue with height never being reported in observationRecord's height coordinate. I haven't yet figured out (or been able to recreate locally) why timeSeries heights are getting reported with profileHeights, still chasing that down...

**Comment by emiliom** [comment]
Great, @shane-axiom! Thanks for including that sample as well.

Let me know if there's something I can do (eg, provide more information) to help you identify what circumstances are associated with profileHeights reporting with a timeSeries.

**Comment by shane-axiom** [comment]
timeSeries profileHeights issue hopefully fixed in 4b31900

**Comment by emiliom** [comment]
Woo-hoo! Thank you.

**Comment by emiliom** [comment]
@shane-axiom, have you thought about making a new release (1.0.1?) to incorporate these bug fixes? I haven't upgraded to 1.0.0, but was planning to do it very soon; now I'll wait until a new release that includes these height-related bug fixes. Thanks.

**Comment by emiliom** [comment]
FYI, I've uploaded [Shane's sample response](https://github.com/ioos/i52n-sos/issues/24#issuecomment-170694620) to my collection of responses from various sos server types and instances at https://github.com/emiliom/stuff/tree/master/ioossosxml. This collection of pre-generated responses will be handy for pyoos testing and fixes.

**Comment by shane-axiom** [comment]
Addressed in 1.1 release:
https://github.com/ioos/i52n-sos/releases/tag/1.1
